### PR TITLE
cql3: Drop unneeded filtering when continuous clustering-key is selected

### PR DESCRIPTION
**Pull request by dekimir**
I noticed that we require filtering for continuous clustering key, which is not necessary.  I dropped the requirement and made sure the correct data is read from the storage proxy.

The corresponding dtest PR: https://github.com/scylladb/scylla-dtest/pull/1727

Tests: unit (dev,debug), dtest (next-gating, cql*py)

### CODE REVIEW

File: test/boost/cql_query_group_test.cc
Comment (avikivity) [inline test/boost/cql_query_group_test.cc:229]: Why is the test changing? Is there a change in behavior?

File: test/boost/cql_query_test.cc
Comment (avikivity) [inline test/boost/cql_query_test.cc:2401]: And why this?

**Comment by dekimir** [comment]
> @dekimir Is the behavior compatible with Apache Cassandra after this change?

No.  Some queries prohibited by Cassandra will now be allowed by Scylla.  For example, Cassandra forbids
```cqlsh
cqlsh:ks> CREATE TABLE t ( p int, c int, r int, primary key(p,c));
cqlsh:ks> SELECT * FROM t WHERE c < 10;
InvalidRequest: Error from server: code=2200 [Invalid query] message="Cannot execute this query as it might involve data filtering and thus may have unpredictable performance. If you want to execute this query despite the performance unpredictability, use ALLOW FILTERING"
```
I don't see why this is necessary; it shouldn't be hard to craft a read command to fetch precisely the slice requested and nothing more (from all shards involved).  All our tests still pass after this change, generating correct query results; this suggests that the filtering turned off by this change wasn't required.

File: test/boost/cql_query_group_test.cc
Comment (avikivity) [reply test/boost/cql_query_group_test.cc:229]: Ah, explained in the changelog.

Comment (dekimir) [reply test/boost/cql_query_group_test.cc:229]: See the commit comment.  There is a change in the behaviour observable by cql_test_env, not by the end user.

File: test/boost/cql_query_test.cc
Comment (avikivity) [reply test/boost/cql_query_test.cc:2401]: ignore

Comment (dekimir) [reply test/boost/cql_query_test.cc:2401]: Explained in the commit message.  The extra values were read to perform filtering, and they're visible to the test.

**Comment by avikivity** [comment]
Looks good, but @psarna should review.

File: cql3/restrictions/statement_restrictions.cc
Comment (avikivity) [inline cql3/restrictions/statement_restrictions.cc:507]: What if the clustering column restrictions are not in order?

`SELECT * FROM tab WHERE ck2 = ?` (ck1 unrestricted)?

Comment (dekimir) [reply cql3/restrictions/statement_restrictions.cc:507]: Handled by line 504.  We have tests for it.

**Comment by psarna** [comment]
This kind of query does not technically involve filtering, but let me show a somehow related issue and @slivne's comment: https://github.com/scylladb/scylla/issues/5547#issuecomment-574160331 . There, filtering is not really applied per se, but it may still make sense to guard this kind of statement with `ALLOW FILTERING` in order to warn the users that it may have performance penalty. In other words, we (and Cassandra) tend to misuse `ALLOW FILTERING` as a more generic `ALLOW POTENTIALLY CRAPPY PERFORMANCE`, and perhaps this case is also worth guarding?
On the other hand, we don't guard `SELECT *` full scans, and this statement is effectively a trimmed full scan...
Hard to decide. I lean towards merging this pull request, because this kind of statement definitely does not need **filtering**. It will potentially take awfully long to execute, but that's no different from a full scan, which also does not require any additional keywords.

File: cql3/restrictions/statement_restrictions.cc
Comment (psarna) [reply cql3/restrictions/statement_restrictions.cc:507]: Well, aside everything I stated in my previous reply, I am definitely in favor of simplifying this abomination of a condition... which is yet another argument for going ahead and merging this patch.

**Comment by avikivity** [comment]
Full scans do not have bad performance. They are slow, but performance per row is very good (much better than single-row queries that miss cache).

Filtering _can_ be bad if selectivity is high - time per returned row is low.

Full scans with slices can be quite efficient.

**Review by psarna — APPROVED** [review]

**Comment by dekimir** [comment]
@psarna @avikivity can one of you please merge?

**Comment by psarna** [comment]
Sure, I can merge tomorrow morning

**Comment by dekimir** [comment]
Please keep in mind that _needs filtering_ is not the same as _needs ALLOW FILTERING_.  The primary goal of this PR is to avoid pointless application of a filtering visitor to each fetched row.

**Comment by nyh** [comment]
Let me offer one possible definition (and ask you to provide a different definition if you don't agree with this one):

**A "filtering" query is a query where the amount of I/O might _not_ be proportional to the amount of output**

In this sense a full table scan is not filtering - while it does a lot of I/O, is also does a lot of output. Reading a slice of a single partition is also not filtering - there is O(1) I/O (not really... but let's pretend....) work to find the beginning of the slice and then you just do I/O for the actual rows you intend to output and not more.

But the "SELECT * FROM t WHERE c < 10;" **is**, unfortunately, filtering under this definition. You may need to do I/O amounting to reading the entire content of the index file, just to generate 10 or even 0 items in the output.

If you have a different definition of what "filtering" is - perhaps one that excludes the index-file I/O (?) - let's formalize this definition and write it somewhere.

**Comment by nyh** [comment]
> Please keep in mind that _needs filtering_ is not the same as _needs ALLOW FILTERING_. The primary goal of this PR is to avoid pointless application of a filtering visitor to each fetched row.

But I saw in the discussion about that one of the results of this patch is that "ALLOW FILTERING" also changed its meaning . If its meaning changed by mistake, I think you should fix it back before committing this patch. If it's not a mistake, we need to understand why it's not a mistake. And for that I need to understand what exactly is "ALLOW FILTERING" supposed to mean.

**Comment by dekimir** [comment]
> If you have a different definition of what "filtering" is - perhaps one that excludes the index-file I/O (?) - let's formalize this definition and write it somewhere.

**I** don't have a definition as much as the Scylla code does:
https://sourcegraph.com/github.com/scylladb/scylla/-/blob/cql3/statements/select_statement.cc#L304:16

Its definition is "a filtering query is a query that will feed all fetched rows through `result_set_builder::restrictions_filter`, one by one.

**Comment by dekimir** [comment]
> > Please keep in mind that _needs filtering_ is not the same as _needs ALLOW FILTERING_. The primary goal of this PR is to avoid pointless application of a filtering visitor to each fetched row.
> 
> But I saw in the discussion about that one of the results of this patch is that "ALLOW FILTERING" also changed its meaning . If its meaning changed by mistake, I think you should fix it back before committing this patch. If it's not a mistake, we need to understand why it's not a mistake. And for that I need to understand what exactly is "ALLOW FILTERING" supposed to mean.

OK, you have a point -- this PR changes both _needs filtering_ and _needs ALLOW FILTERING_.  If @avikivity agrees that the _needs ALLOW FILTERING_ behaviour must not be disturbed, I'm happy to change it back to what it was.

**Comment by nyh** [comment]
> > If you have a different definition of what "filtering" is - perhaps one that excludes the index-file I/O (?) - let's formalize this definition and write it somewhere.
> 
> **I** don't have a definition as much as the Scylla code does:
> https://sourcegraph.com/github.com/scylladb/scylla/-/blob/cql3/statements/select_statement.cc#L304:16
> 
> Its definition is "a filtering query is a query that will feed all fetched rows through `result_set_builder::restrictions_filter`, one by one.

I don't understand. Is this a definition of needs_filtering (some internal implementation detail) or of "ALLOW FILTERING" (a user facing feature)?

Also, why is it ok for this definition to only mention rows, not partitions? What about the problem that sometimes, deciding which rows to fetch - from a much larger set of partitions - is itself is a type of filtering that the user might also want to avoid?

**Comment by dekimir** [comment]
> But the "SELECT * FROM t WHERE c < 10;" **is**, unfortunately, filtering under this definition. You may need to do I/O amounting to reading the entire content of the index file, just to generate 10 or even 0 items in the output.

Would it be any different if you added `AND token(p) > 123`?  That currently doesn't require ALLOW FILTERING but I think it would require it under your definition.

**Comment by dekimir** [comment]
> > > If you have a different definition of what "filtering" is - perhaps one that excludes the index-file I/O (?) - let's formalize this definition and write it somewhere.
> > 
> > 
> > **I** don't have a definition as much as the Scylla code does:
> > https://sourcegraph.com/github.com/scylladb/scylla/-/blob/cql3/statements/select_statement.cc#L304:16
> > Its definition is "a filtering query is a query that will feed all fetched rows through `result_set_builder::restrictions_filter`, one by one.
> 
> I don't understand. Is this a definition of needs_filtering (some internal implementation detail) or of "ALLOW FILTERING" (a user facing feature)?

The former.

> Also, why is it ok for this definition to only mention rows, not partitions?

Because partitions are an implementation detail from this definition's perspective.  Either the fetched rows are precisely what the query asked for or they are a superset.

> What about the problem that sometimes, deciding which rows to fetch - from a much larger set of partitions - is itself is a type of filtering that the user might also want to avoid?

That's a separate question from this `need_filtering` method.  It has to do with user's expectations.

**Comment by nyh** [comment]
> > But the "SELECT * FROM t WHERE c < 10;" **is**, unfortunately, filtering under this definition. You may need to do I/O amounting to reading the entire content of the index file, just to generate 10 or even 0 items in the output.
> 
> Would it be any different if you added `AND token(p) > 123`? That currently doesn't require ALLOW FILTERING but I think it would require it under your definition.

In theory, "token(p) > 123" just requires us to do one read in the index file - to find token 123 - and skip the right position in the data file(s), and start reading sequentially from there - outputting whatever we read. If this is what we do, there is no "filtering". But I don't know if this is really how we implemented it.

**Comment by psarna** [comment]
Well, if there's a disagreement whether `ALLOW FILTERING` should stay or not, the safer option is for it to stay, just for the sake of compatibility with Cassandra if nothing else. I'm personally in favor of simplifying the rules for `ALLOW FILTERING`, but I don't have any strong arguments - the only one is that `ALLOW FILTERING` was originally meant to be used for queries which require coordinator-side filtering, and this one is not the case. But that's just arguing about definitions of `ALLOW FILTERING`, and I'm fine with accepting another definition if it makes other people happy (:

**Comment by avikivity** [comment]
> Well, if there's a disagreement whether `ALLOW FILTERING` should stay or not, the safer option is for it to stay, just for the sake of compatibility with Cassandra if nothing else. I'm personally in favor of simplifying the rules for `ALLOW FILTERING`, but I don't have any strong arguments - the only one is that `ALLOW FILTERING` was originally meant to be used for queries which require coordinator-side filtering, and this one is not the case. But that's just arguing about definitions of `ALLOW FILTERING`, and I'm fine with accepting another definition if it makes other people happy (:

I think we should drop the requirement for ALLOW FILTERING, if we don't need to filter. It's needlessly scares people into thinking their query is bad/slow.

**Comment by nyh** [comment]
> I think we should drop the requirement for ALLOW FILTERING, if we don't need to filter. It's needlessly scares people into thinking their query is bad/slow.

@avikivity in the specific example `SELECT * FROM t WHERE c < 10;` isn't it bad/slow? It may generate a couple of results, and yet need to scan the entire index file (at least) to do it. I guess it's not the "worst" example of filtering, but isn't it still "bad"?

**Comment by dekimir** [comment]
> > > But the "SELECT * FROM t WHERE c < 10;" **is**, unfortunately, filtering under this definition. You may need to do I/O amounting to reading the entire content of the index file, just to generate 10 or even 0 items in the output.
> > 
> > 
> > Would it be any different if you added `AND token(p) > 123`? That currently doesn't require ALLOW FILTERING but I think it would require it under your definition.
> 
> In theory, "token(p) > 123" just requires us to do one read in the index file - to find token 123 - and skip the right position in the data file(s), and start reading sequentially from there - outputting whatever we read.

Just to be clear, I'm asking about these two queries:
1. `WHERE c < 10`
2. `WHERE c < 10 AND token(p) > 123`

Let's say, for the sake of argument, that every partition's token is over 123.  I don't see how query 2 results in any less IO than query 1.  If I'm right, then they should both require ALLOW FILTERING.  But query 2 doesn't require it today.

**Comment by avikivity** [comment]
> > I think we should drop the requirement for ALLOW FILTERING, if we don't need to filter. It's needlessly scares people into thinking their query is bad/slow.
> 
> @avikivity in the specific example `SELECT * FROM t WHERE c < 10;` isn't it bad/slow? It may generate a couple of results, and yet need to scan the entire index file (at least) to do it. I guess it's not the "worst" example of filtering, but isn't it still "bad"?

No, it needs to binary search c=10 in order to find the edge.

**Comment by dekimir** [comment]
@avikivity if you're happy with the PR as it stands, please merge it.  If not, please explain what needs to change.

**Comment by avikivity** [comment]
I'm happy with it, and @psarna approves, but I'd like to achieve consensus with @nyh.

**Comment by psarna** [comment]
Well, today I learned that we may need a couple of postmen and a lake for that.

**Comment by dekimir** [comment]
Ping.  This is blocking subsequent development.

@nyh please address these two comments:
https://github.com/scylladb/scylla/pull/7460#issuecomment-717891414
https://github.com/scylladb/scylla/pull/7460#issuecomment-717898327

**Comment by nyh** [comment]
> > > I think we should drop the requirement for ALLOW FILTERING, if we don't need to filter. It's needlessly scares people into thinking their query is bad/slow.
> > 
> > 
> > @avikivity in the specific example `SELECT * FROM t WHERE c < 10;` isn't it bad/slow? It may generate a couple of results, and yet need to scan the entire index file (at least) to do it. I guess it's not the "worst" example of filtering, but isn't it still "bad"?
> 
> No, it needs to binary search c=10 in order to find the edge.

"c" in my example was a clustering key. Every single partition needs to be search for this c=10. There maybe a million of them, but only one or zero results. This is why I call it filtering.

**Comment by nyh** [comment]
> > > > But the "SELECT * FROM t WHERE c < 10;" **is**, unfortunately, filtering under this definition. You may need to do I/O amounting to reading the entire content of the index file, just to generate 10 or even 0 items in the output.
> > > 
> > > 
> > > Would it be any different if you added `AND token(p) > 123`? That currently doesn't require ALLOW FILTERING but I think it would require it under your definition.
> > 
> > 
> > In theory, "token(p) > 123" just requires us to do one read in the index file - to find token 123 - and skip the right position in the data file(s), and start reading sequentially from there - outputting whatever we read.
> 
> Just to be clear, I'm asking about these two queries:
> 
>     1. `WHERE c < 10`
> 
>     2. `WHERE c < 10 AND token(p) > 123`
> 
> 
> Let's say, for the sake of argument, that every partition's token is over 123. I don't see how query 2 results in any less IO than query 1. If I'm right, then they should both require ALLOW FILTERING. But query 2 doesn't require it today.

@dekimir ok, you convinced me. That "ALLOW FILTERING" is ill-defined before this patch, and continues to be ill-defined after this patch, so that's ok, even if a bit disappointing. I was hoping we could have a clear definition what "ALLOW FILTERING" should mean, and base the implementation on it, and not the opposite.

Anyway, I'll re-review and most likely commit, as I see both @avikivity and @psarna approved already.

File: test/boost/cql_query_test.cc
Comment (nyh) [reply test/boost/cql_query_test.cc:2401]: This is https://github.com/scylladb/scylla/issues/4126. Interestingly, this extra values was the main reason why I lost faith in doing CQL tests via some invented interface (be it cql_tests.cc as here or cql_repl) and think we should be doing CQL tests via a real CQL driver.

**Review by nyh — APPROVED** [review]
I'm not happy that we are changing the implementation of "ALLOW FILTERING" without first defining anywhere what "ALLOW FILTERING" should mean, or even deciding if it needs to be compatible with Cassandra. On the other hand, but I also don't think there's any error in this per se - it's not less "correct" than it was, I think it some cases it saves useless work, and two other maintainers already approved it, so I'll approve too and commit.

**Comment by avikivity** [comment]
Fails dtest, dequeueing:

```
04:17:14  ======================================================================
04:17:14  FAIL: allow_filtering_test (cql_additional_tests.TestCQL)
04:17:14  ----------------------------------------------------------------------
04:17:14  Traceback (most recent call last):
04:17:14    File "/jenkins/workspace/scylla-master/next/scylla-dtest/cql_additional_tests.py", line 3076, in allow_filtering_test
04:17:14      self._assert_invalid_filtering(session=session, query=q)
04:17:14    File "/jenkins/workspace/scylla-master/next/scylla-dtest/cql_additional_tests.py", line 6188, in _assert_invalid_filtering
04:17:14      assert_invalid_case_insensitive_matching(session=session, query=query, matching=MSG_ALLOW_FILTERING)
04:17:14    File "/jenkins/workspace/scylla-master/next/scylla-dtest/assertions.py", line 28, in assert_invalid_case_insensitive_matching
04:17:14      raise e
04:17:14    File "/jenkins/workspace/scylla-master/next/scylla-dtest/assertions.py", line 26, in assert_invalid_case_insensitive_matching
04:17:14      assert False, "Expecting query to be invalid: got %s" % res
04:17:14  AssertionError: Expecting query to be invalid: got <cassandra.cluster.ResultSet object at 0x7f1a63856750>
```

**Comment by avikivity** [comment]
Actually, looks like it was already dequeued by someone else.

**Comment by dekimir** [comment]
@avikivity @penberg dtest has been [changed](https://github.com/scylladb/scylla-dtest/commit/67260d1a581de59f79dc998265e488241202a98d) not to fail, so it's safe to merge this now.

**Pull request closed without merging**
